test/topology_experimental_raft/test_alternator.py

```diff
@@ -318,9 +318,85 @@ async def check_two_live_nodes():
+    """A test for /localnosts on a more general setup, with multiple DCs and
+       multiple racks - an 8-node setup with two DCs, two racks in each, and
+       two nodes in each rack.
+       Test both the default of returning the nodes on DC of the server being
+       connected - and the "dc" and "rack" options for explicitly choosing a
+       specific dc and/or rack.
+    """
+    # Start 8 nodes on two different dcs (called "dc1" and "dc2") and two
+    # different rack ("rack1" and "rack2"), two nodes in each rack.
+    config = alternator_config | {
+        'endpoint_snitch': 'GossipingPropertyFileSnitch'
+    }
+    servers = {}
+    for dc in ['dc1', 'dc2']:
+        for rack in ['rack1', 'rack2']:
+            servers[dc,rack] = await manager.servers_add(2, config=config, property_file={
+                'dc': dc, 'rack': rack})
+
+    def localnodes_request(server):
+        return f"http://{server.ip_addr}:{alternator_config['alternator_port']}/localnodes"
+
+    # Before we test various variations of the /localnodes request, let's wait
+    # until all nodes are visible to each other in /localnodes requests. This
+    # can take time, while nodes finish bootstrapping and gossip to each other
+    # (see #19694). After this one-time wait_for, the following checks will be
+    # able to check things immediately - without retries.
+    for dc in ['dc1', 'dc2']:
+        for rack in ['rack1', 'rack2']:
+            for server in servers[dc, rack]:
+                async def check_localnodes_four():
+                    response = requests.get(localnodes_request(server))
+                    if len(json.loads(response.content.decode('utf-8'))) == 4:
+                        return True
+                    return None # try again
+                assert await wait_for(check_localnodes_four, time.time() + 60)
+
+    # Check that the option-less "/localnodes" returns for each of dc1's nodes
+    # the four dc1 servers, and for each of dc2's nodes, the four dc2 servers:
+    for dc in ['dc1', 'dc2']:
+        dc_servers = servers[dc, 'rack1'] + servers[dc, 'rack2']
+        expected_ips = [server.ip_addr for server in dc_servers]
+        for server in dc_servers:
+            response = requests.get(localnodes_request(server))
+            assert sorted(json.loads(response.content.decode('utf-8'))) == sorted(expected_ips)
+
+    # Check that the "dc" option works - it should return the nodes for the
+    # specified DC, regarless of which node on which DC the request is sent
+    # to (we test all combinations of one of 8 target nodes and 2 option dcs).
+    all_servers = sum(servers.values(), [])
+    for option_dc in ['dc1', 'dc2']:
+        expected_servers = servers[option_dc, 'rack1'] + servers[option_dc, 'rack2']
+        expected_ips = [server.ip_addr for server in expected_servers]
+        for server in all_servers:
+            response = requests.get(localnodes_request(server)+f'?dc={option_dc}')
```

Comment:
Maybe `localnodes_request` function could just take optional parameters: dc & rack, instead of us concatenating the string like here, this would perhaps be more readable.

Reply:
Would replacing "+" in this line by "," make this line more readable?
Or do you mean using `**kwargs` (which then introduces the join logic into this tiny localnodes_request function - is it worth it?)

Reply:
No, I just meant to add `dc` and `rack` parameters to the `localnodes_request` and if any is present, you just append e.g. "?dc={dc}" to the URL string - does this make sense? If not, you can ignore it, it's not a big deal after all.

Reply:
But I use this function just a handful of times in this test, and each call has a different combination of these options. So either I make it `**kwargs` and concatenate the keywords that I get (different ones in each call) or leave it the way it is. Adding two explicit parameters dc and rack will force me to add ugly if's to this one-line function.

Reply:
So `localnodes_request` is a single line function that just returns a parameterized string. (If python had C++ semantics, this could be replaced with a variable template and wouldn't require a function). I'm not sure why to wrap exacly this piece of URL into a string, and not the whole URL, as you said it's perhaps pragmatic; currently the testcode mixes levels of abstractions: it hides some implementations details under a function but still exposes these details in the body of a testcase, which ideally should just give a high-lvl overview of what's being tested, and this would be achieved if concatenation of URL was extracted into a separate function. Other than that `localnodes_request`'s name seem to indicate it returns, well, localnodes request, while it returns half of it? But also parameterized by the serwer? And not by other parameters?

> will force me to add ugly...

well, for the reasons listed above, I'm not sure if currently the code is all that beautiful ;)

> ... if's to this one-line function

how so, what I had in mind is something along the lines of (warning, pseudocode):
```
    def localnodes_request(server, dc, rack):
        return f"http://{server.ip_addr}:{alternator_config['alternator_port']}/localnodes?" + (dc if dc else "") + "&" + (rack if rack else "")
```
(I think both ? and & won't hurt the URL, even if both new params are empty)
Then you call it however you like, e.g.:
```suggestion
            response = requests.get(localnodes_request(server, dc:option_dc)
```

Reply:
The reason why I created this function is simple - it's a long and ugly string needed exactly in the same way 5 times in the code. And it's not just a template, it actually needs to extract server.ip_addr. It returns a URL to which the 5 different calls need to tack on 5 different sets of parameters.

I'll rethink it, but I think we're over-designing this. My main goal was for a future reader to see the call:
```
response = requests.get(localnodes_request(server)+f'?rack={option_rack}')
```
and immediately understand what is going on here... There is the basic localnodes GET request URL, to which you need to tack on the options string, like `?rack=something&dc=something`. It should be 100% self-explanatory to anyone familiar with HTTP. I didn't even think it's confusing.

You're right that
```
    def localnodes_request(server, dc, rack):
        return f"http://{server.ip_addr}:{alternator_config['alternator_port']}/localnodes?" + (dc if dc else "") + "&" + (rack if rack else "")
```

Would work (and there's no need for the "if else" - python also has defaults for parameters) - but has a big flaw: when dc isn't passed, it still passes "dc=". This will work (I know, I wrote the code that reads it and "dc=" behaves exactly the same as a missing "dc"), but I really wanted to check that the code still works without dc passed at all.

Reply:
I'm changing this in the next version